python/lsst/pipe/base/_status.py

```diff
@@ -73,6 +81,108 @@ class RepeatableQuantumError(RuntimeError):
+    processed, even by a better algorithm.
+    """
+
+
+class AnnotatedPartialOutputsError(RepeatableQuantumError):
+    """Exception that runQuantum raises when the (partial) outputs it has
+    written contain information about their own incompleteness or degraded
+    quality.
+
+    This exception should always chain the original error. When the
+    executor catches this exception, it will report the original exception. In
+    contrast, other exceptions raised from ``runQuantum`` are considered to
+    invalidate any outputs that are already written.
+    """
+
+    @classmethod
+    def annotate(cls, exception: Exception, *, task, exposures=(), catalogs=()) -> None:
+        """Set metadata on outputs to explain the nature of the failure.
+
+        Parameters
+        ----------
+        exception : `Exception`
+            Exception that caused the task to fail.
+        task : `lsst.pipe.base.Task`
+            Task that failed and that will be annotated.
+        exposures : `tuple` [`lsst.afw.image.Exposure`], optional
+            Exposures to annotate with information about the failure.
+        catalogs : `tuple` [`lsst.afw.table.Catalog`], optional
+            Catalogs to annotate with information about the failure.
+
+        Returns
+        -------
+        error : `AnnotatedPartialOutputsError`
+            Exception that the failing task can ``raise from`` with the
+            passed-in exception.
+
+        Notes
+        -----
+        This should be called from within an except block that has caught an
+        exception. Here is an example of handling a failure in
+        ``PipelineTask.runQuantum`` that annotates and writes partial outputs
+        outputs:
+
+        .. code-block:: py
+            :name: annotate-error-example
+
+            def runQuantum(self, butlerQC, inputRefs, outputRefs):
+                inputs = butlerQC.get(inputRefs)
+                exposures = inputs.pop("exposures")
+                assert not inputs, "runQuantum got more inputs than expected"
+
+                result = pipeBase.Struct(catalog=None)
+                try:
+                    self.run(exposure)
+                except pipeBase.RepeatableQuantumError as e:
+                    error = pipeBase.AnnotatedPartialOutputsError.annotate(
+                        e,
+                        catalogs=[result.catalog],
+                        task=self)
+                    raise error from e
+                finally:
+                    butlerQC.put(result, outputRefs)
+        """
+        # NOTE: Can't test this in pipe_base because afw is not a dependency;
+        # test_calibrateImage.py in pipe_tasks provides significant coverage.
+        for item in itertools.chain(exposures, catalogs):
+            # Some outputs may not exist, so we cannot set metadata on them.
+            if item is None:
+                continue
+            if (metadata := item.getMetadata()) is None:
```

Comment:
My dream would be to just pass the result `Struct` (or maybe `result.getDict()`?) into this `annotate` and let it annotate all the non-None items. That's probably wishful thinking though.

Part of me is tempted to put an `if isinstance(item, astropy.table.Table): item.meta.update(metadata)`  to get that...

Reply:
I'm fine with having the `SourceTable` constructor initialize a `PropertyList` (it'll be `SourceTable` rather than `SourceCatalog` you want to modify, because it's `BaseTable` that holds the actual `shared_ptr<PropertyList>`.

Having a `meta` alias on our objects doesn't actually bring us enough closer to `astropy.table.Table` because I'm sure their `meta` object does not have the `set_dict` and `get_dict` methods I added to `PropertyList`, `PropertySet`, and `TaskMetadata` on the other ticket.

Maybe an adapter class and context manager would work:
```
@dataclasses.dataclass
class GetSetDictMetadataAdapter:
    metadata: TaskMetadata

@contextlib.contextmanager
def adapted_astropy_metadata(self, table: astropy.table.Table):
    metadata = TaskMetadata()
    yield GetSetDictMetadataAdapter(metadata)
    table.meta.update(metadata.to_dict())
```

That doesn't necessarily get you closer to being able to pass in just the struct to `annotate`, but I don't think we want `annotate` to have to do runtime checking to see if each item in the struct actually can hold metadata anyway.  I like making what's annotated explicit in the task implementation, too.